File: hour_glass_sum_hacker_rank.go
package main

import (
	"fmt"
	"math"
)

// https://www.hackerrank.com/challenges/2d-array/problem
func hourglassSum(arr [6][6]int32) (maxHourGlassSum int32) {
	for rowIndex := range arr {
		for columnIndex := range arr[0] {
			// 3 rows of overhead needed for 3 levels of the hourglass (top, middle, bottm)
			// 3 columns of overhead needed for length of top and bottom of hour glass
			if rowIndex > 3 || columnIndex > 3 {
				break
			}
			fmt.Printf("row %d  col %d\n", rowIndex, columnIndex)
			topHourGlass := arr[rowIndex][columnIndex] + arr[rowIndex][columnIndex+1] + arr[rowIndex][columnIndex+2]
			HourGlassBar := arr[rowIndex+1][columnIndex+1]
			bottomHourGlass := arr[rowIndex+2][columnIndex] + arr[rowIndex+2][columnIndex+1] + arr[rowIndex+2][columnIndex+2]

			hourGlassSum := topHourGlass + HourGlassBar + bottomHourGlass
			// initial condition for max sum
			if columnIndex == 0 && rowIndex == 0 {
				maxHourGlassSum = hourGlassSum
			}
			maxHourGlassSum = int32(math.Max(float64(maxHourGlassSum), float64(hourGlassSum)))
		}
	}
	return
}

func main() {
	arr := [6][6]int32{
		{1, 1, 1, 0, 0, 0},
		{0, 1, 0, 0, 0, 0},
		{1, 1, 1, 0, 0, 0},
		{0, 0, 2, 4, 4, 0},
		{0, 0, 0, 2, 0, 0},
		{0, 0, 1, 2, 4, 0},
	}
	fmt.Println(arr)
	sum := hourglassSum(arr)
	fmt.Println(sum)
}

File: hour_glass_sum_hacker_rank.py
#!/bin/python3
# https://www.hackerrank.com/challenges/2d-array/problem

import math
import os
import random
import re
import sys

# Complete the hourglassSum function below.
def hourglassSum(arr):
    for row in range(len(arr)):
        for col in range(len(arr[0])):
            """
            arr[0][0]+arr[0][1]+arr[0][2]+arr[1][1]+arr[2][0]+arr[2][1]+arr[2][2]
            arr[0][1]+arr[0][2]+arr[0][3]+arr[1][2]+arr[2][1]+arr[2][2]+arr[2][3]
            arr[0][2]+arr[0][3]+arr[0][4]+arr[1][3]+arr[2][2]+arr[2][3]+arr[2][4]
            arr[0][3]+arr[0][4]+arr[0][5]+arr[1][4]+arr[2][3]+arr[2][4]+arr[2][5]

            arr[1][0]+arr[1][1]+arr[1][2]+arr[2][1]+arr[3][0]+arr[3][1]+arr[3][2]
            ...
            ..
            arr[1][3]+arr[1][4]+arr[1][5]+arr[2][4]+arr[3][3]+arr[3][4]+arr[3][5]
            ###

            arr[3][3]+arr[3][4]+arr[3][5]+arr[4][4]+arr[5][3]+arr[5][4]+arr[5][5]
            """
            if col > 3 or row > 3:
                break
            hour_glass_sum = (
                arr[row][col]
                + arr[row][col + 1]
                + arr[row][col + 2]
                + arr[row + 1][col + 1]
                + arr[row + 2][col]
                + arr[row + 2][col + 1]
                + arr[row + 2][col + 2]
            )
            if col == 0 and row == 0:  # initial condition
                max_hour_glass_sum = hour_glass_sum
            max_hour_glass_sum = max(max_hour_glass_sum, hour_glass_sum)
    return max_hour_glass_sum


if __name__ == "__main__":

    arr = [
        [1, 1, 1, 0, 0, 0],
        [0, 1, 0, 0, 0, 0],
        [1, 1, 1, 0, 0, 0],
        [0, 0, 2, 4, 4, 0],
        [0, 0, 0, 2, 0, 0],
        [0, 0, 1, 2, 4, 0],
    ]
    result = hourglassSum(arr)
    print(result)
